### ISSUE TITLE
Support an `"engines"` field in `package.json`

### ISSUE THREAD

**Issue by lgarron** [oven-sh/bun#5846]
### What is the problem this feature would solve?

In https://github.com/oven-sh/bun/issues/5845 I was testing with an outdated version of `bun`. Although I was creating a project from scratch, it is possible to catch similar compatibility issues *among developers of a given repo* by using the [`"engines"` field](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines) of `package.json` — in particular, `bun` could show a compatibility warning (optionally overridable) instead of erroring deep in implementation code.

For those of us testing various versions of `bun`, this could be particularly useful to avoid head-scratching situations. It's also just generally useful for any project that is using features of `bun` as they become available, if any of those projects are checked out by multiple developers or on multiple devices.

### What is the feature you are proposing to solve the problem?

Support `"bun"` version ranges in the `"engines"` field of `package.json` for both the project and dependencies and fail `bun` commands (particularly `bun x` and `bun run`) by default if the current `bun` version does not match the constraint.

### What alternatives have you considered?

Leave things as-is, especially if it's a performance concern?

**Comment by paperclover** [comment]
i dont think its a huge big deal, but right now yarn puts a warning for packages that use an unknonw engine. but this is the only package manager to do it.

![image](https://github.com/oven-sh/bun/assets/24465214/6e04843a-635c-4e47-8da0-fe491a0a8930)

otherwise, this is something we've thought about.

**Comment by adematte** [comment]
as most teams will end up with a mix of node and bun for their projects, this would make sense to support the `engines` field of package.json. Otherwise we are left with only README.md to indicate what version of bun or node (or both), we need to run a given project.

**Comment by Kleywalker** [comment]
Are there any plans to implement this feature?

**Comment by paperclover** [comment]
yes but probably not right now. i think it would be cool if bun in the engines field could be used to force the `--bun` flag for clis

**Comment by gornostay25** [comment]
@lgarron Any updates?

**Comment by strafe** [comment]
Seems like this is the only reasonable way to know which runtime a project is meant to be used with. For reproducibility's sake, this is kind of needed. Once this is done and bun is supported by corepack we can all sleep easy.

**Comment by Kleywalker** [comment]
Is this being considered as a possible feature, and if so, will it be implemented in the near future?

**Comment by afonsojramos** [comment]
> I came across this while trying to find a way to force all users of our monorepo to update to the newly released `bun@1.1.9`.
> 
> It looks like the `oven-sh/setup-bun@v1` GitHub action [searches `package.json` for the string `bun@`](https://github.com/oven-sh/setup-bun/blob/main/src/index.ts?rgh-link-date=2024-05-23T10%3A32%3A13Z) in the `packageManager` field which is what [corepack](https://nodejs.org/api/corepack.html) would use if corepack supported bun.
> 
> Is there any way to try and force all users to use a particular version of bun or should I add a test which checks `Bun.version` against `package.json::packageManager` or something?

@vitch Nice callout!

Found this in the `setup-bun` action, pay attention to the bum file, which references: https://github.com/owenizedd/bum
<img width="596" alt="image" src="https://github.com/oven-sh/bun/assets/19473034/813efdad-0ba6-483c-b510-9aabc3c7393e">

Might be a good intermediate solution while bun itself doesn't add support to it.

> Note: I'll try to setup the bun project locally today and try to work on this. Can't guarantee success as it will be the first dabble in Zig

**Comment by therealparmesh** [comment]
Relates: It looks like [corepack is being removed from the node distribution](https://github.com/nodejs/package-maintenance/pull/606).

**Comment by wesleytodd** [comment]
I saw this issue because of the link in the comment above. In case folks hadn't seen this, a group of folks have been working on a more robust proposal to cover `packageManager` and related things. We are calling it `devEngines` and it looks to be acceptable by both the `npm` and `pnpm` teams.  Worth considering in this conversation I would think.

https://github.com/openjs-foundation/package-metadata-interoperability-collab-space/issues/15

**Comment by lgarron** [comment]
Now that Bun is gaining a new lockfile format, I think it would be more valuable than ever to be able to specify "this project requires a minimum `bun` version".

**Comment by dy0gu** [comment]
Any plans for this? It would aid migration to the new lockfile in codebases with multiple devs that may not have bun updated.

**Comment by vitch** [comment]
FWIW I just wrote a `check-bun-version.ts` script which looks like this:

```ts
#!/usr/bin/env bun
let { default: packageJson } = await import('../../package.json');

let { packageManager } = packageJson;
let [, expectedBunVersion] = packageManager.split('@');

if (expectedBunVersion === Bun.version) {
  console.log(`Bun version ${expectedBunVersion} ok`);
} else {
  throw new Error(
    `Running with wrong bun version. Got ${Bun.version}, expected ${expectedBunVersion}`,
  );
}
```

I call the script at the head of my other scripts. It's not fool proof but it has helped keep our small team synced and make sure no one forgets to update. And it's reading the same field that CI does (via `oven-sh/setup-bun`) which means we're synced with that too...

Something built in would obviously be  way better but this might help someone out in the meantime...

**Comment by hadarziv-army** [comment]
Is there an ETA for this issue?

**Comment by mountainash** [comment]
@adamhl8 cool idea, but as package.json can be read (like a normal .json) you could have the version requirement in there under the “engines” key - so it’s not hardcoded in some other file and all your requirements are together. It should also be backwards compatible for when/if official support comes.

**Comment by Jarred-Sumner** [comment]
I've opened a PR to yarn v1 to no longer warn when `"bun"` is used in `"engines"`

https://github.com/yarnpkg/yarn/pull/9214
